### PR TITLE
[TODO] Migration scripts odoo-brazil to oca

### PR DESCRIPTION
**Pull request by mileo**
Hi,

I performed a database analysis using https://github.com/OCA/OpenUpgrade

As i said in other topics we have this migrations issues, because the migration divergences with @akretion and other contributors with this fork: https://github.com/odoo-brazil/l10n-brazil/

And in the end of the year we have a big issue, with the amendments as are required by the government in product invoicing:

Nota Técnica 2015.003 - v 1.40: ICMS em Operações Interestaduais de Vendas a Consumidor Final
http://www.nfe.fazenda.gov.br/portal/exibirArquivo.aspx?conteudo=yldA7bYcnVg=

I will work in the next days NT2015/003, I request that PSCs can work in this migration to don't stop product invoicing in any Community Odoo user.

Best regards

### CODE REVIEW

**Comment by mileo** [comment]
I forgot the report:

---Fields in module 'general'---
- 4036 fields matched,
- Direct match: 4015
- Found in other module: 6
- Found with different name: 0
- Found with different type: 0
- In obsolete models: 15
- New columns: 119
- Not matched: 22

new model l10n_br_account_product.icms_relief
new model l10n_br_account_product.ipi_guideline
new model l10n_br_tax.estimate
new model l10n_br_tax.estimate.template
obsolete model l10n_br_tax.definition
obsolete model l10n_br_tax.definition.template
---XML records in module 'general'---
ERROR: module not in list of installed modules:
---Fields in module 'l10n_br_product'---
---XML records in module 'l10n_br_product'---
DEL product.pricelist: product.list0

**Comment by rvalyi** [comment]
I think this analysis will be useful for the people using the the temporary branch you mentioned indeed. Now eventually the safest regarding the end of the year might be to backport the NT 2015.003 to that temporary branch if migrating to the OCA branch is too much of a burden for now (I think you started that work, no?)

A small portion of the change in the OpenUpgrade Analysis is the change to accommodate with upstream refactor by Sylvain Legal in the OCA account_product_fiscal_classification and that is what we annotated here https://github.com/OCA/l10n-brazil/blob/8.0/l10n_br_account_product/migrations/8.0.2.0.0/pre-migration.py#L20
in the OCA branch we did the refactor between versions 8.0.1.0.0 and 8.0.2.0.0. Now when somebody translates these notes into a migration script we can absolutely merge that script in the OCA branch.

We have similar issues at Akretion migrating from our work branches to the OCA branch too, though probably simpler in our case. Renato and me will keep working on merging the last modules in the 8.0 OCA branch and ensuring they have a good test coverage. Many thanks for your work on the NT BTW, that's really a useful contrib and hopefully we can merge it soon (also when you feel it's more in shape with the refactors you announced).

In any case it's important to educate the community about the challenges of implementing Odoo in Brazil and the importance of a contract support unless they are very skilled in the field and ready to invest a significant amount of time. Sadly this is not the sexiest message to sell the project :-(

Hopefully the rate of major changes in Odoo will decrease and our localization will be able to stabilize as well.

**Comment by mileo** [comment]
We just backport https://github.com/odoo-brazil/l10n-brazil/pull/138 and https://github.com/odoo-brazil/l10n-brazil/pull/136

This is the analysis of https://github.com/odoo-brazil/account-fiscal-rule/blob/8.0.2.0/account_product_fiscal_classification/migrations/8.0.2.1.0/openupgrade_analysis.txt

I think that the backport quantity of work will be similar to the migration, backport all this is code is a waste of time.

There are a lot of new feature in the PR of OCA/l10n-brazil, we have to arrive at an end of this "temporary branch", perhaps delete it. ( Remember this temporary branch is held by @akretion too. ) 

In my opinion it is the responsibility of @akretion make this migration script. I'm very busy working with NT2015/003, CNAB, nfe ( issues ) and edoc workflow. If you are not taking responsibility for it let me know.

**Comment by rvalyi** [comment]
@mileo I think we can assume this part https://github.com/OCA/l10n-brazil/blob/8.0/l10n_br_account_product/migrations/8.0.2.0.0/pre-migration.py#L20

As for the other parts well everybody is responsible for what he sold to their customers or said to the community. That's not us who promoted this temp branch and I think some people waited a bit too much before making the switch. We made it clear we were using a branch made of the OCA PR's and not the branch you mentioned and that was working well enough for us to be able to do the SPED, that is quite good since many months (though not perfect yet). We could have merge altogether like you did in the odoo-brazil branch but that would have been no help to meet the OCA standards that's why we kept working on the PR's and avoided telling anybody things would be easy. In any case we already have migration work to migrate to the OCA branch at all our customers. And this happens in many OCA projects when we need something before it's merged in the OCA, not just l10n-brazil.

I know this has been annoying but bare in mind the situation was quite worst on our side and that wasn't by choice. If today I don't want to wait for C2C Magento connector migration funding and quickly do some branch on my side with a couple of other folks I would probably end up with the same migration situation (some people did already in previous versions). So simply I don't try that and I plan my projetcts in accordingly instead. Again that was unfortunate force majeur and fortunately people don't use to die this way every year so there is no chance this occurs again.

**Comment by danimaribeiro** [comment]
So this means you are working alone Raphael?
You just accept help, but are not willing to help the others?

**Comment by rvalyi** [comment]
@danimaribeiro look how many commits we did in the project... Think what will be the ecosystem if we did didn't create the OCA even before you started working on Odoo... Would Luis be able to contribute so much to the project if we didn't invest time training him (more than a week for free)...

We are not responsible if you or somebody else tell people to use some random branch you do because you don't want to wait the OCA merges (and helped little to make it happen soon enough). We are responsible for what we tell people or our customers. I think this is normal, Open source doesn't mean free support for everybody right.

You now seems to make a little drama with the NT. But earlier in the year there was the "lei da transparencia" and that was Renato who did the work and that was a bit similar in term of having to cope with this new fiscal requirement, right...

The beginning of 2015 was bad for a good reason sadly. But the project has 21 contributors. Many coming from the community we spent years building so no I'm sorry we are certainly not working alone as you say. Now, We don't fool ourselves: 95% of the work is done by 3 or 4 people and the time you spend to get new comers on board hardly compensate it. Still we did it a lot. Again don't judge by the beginning of 2015, but rather judge the 6 years building the project.

If you look for instance at the French localization (the major Odoo market), the situation is not so much better in term of how many people do the work: https://github.com/OCA/l10n-france/graphs/contributors

**Comment by rvalyi** [comment]
@danimaribeiro @mileo saying we don't help is a bit of a stretch: @renatonlima even spent time analyzing all commits that were done in the temp branch you used and referenced all the features that might make sense to port in issues such as https://github.com/OCA/l10n-brazil/issues/190 (one by module).

In any case following upstream Odoo change is hard. In v10 the old API will be deprecated, so we had to bite the bullet of the API change, the earlier the better. At least we do warn the community that this isn't easy.

**Comment by rvalyi** [comment]
with such a reasoning you could say that we also need the NT for the folks in the 7.0 branch and that Akretion should provide a script to migrate from 7.0 to 8.0. Sorry but this is flawed.

**Comment by danimaribeiro** [comment]
"A small portion of the change in the OpenUpgrade Analysis is the change to accommodate with upstream refactor by Sylvain Legal in the OCA account_product_fiscal_classification and that is what we annotated here"

Why just this change is important?
Just because the code from akretion need this change?

No problem if you don't want to accept the migration scripts from odoo-brazil to OCA, we are not going to ask anymore. But once again we are seeing how you drive this project.

**Comment by rvalyi** [comment]
@danimar, we do our best. And with this we did may be 20 or 30x more work in this project than you did. And more than 10x-15x what @mileo did. Not because we are smarter but because we started 4 years before. We certainly value your contributions, but sorry we don't accept these cheap critics. Again my wife died in my arms in November 2014. If you cannot understand we got a bit low during several months because of that, sorry but you are not "helping". And still, we overcame that and we are still the ones who do order of magnitude more work in the projects and in the OCA more widely.

Are you doing better? So basically you took the initiative to recommend people to follow some fork of your own because you didn't want to wait or tell people to wait. Now these people are in trouble with this un-maintained adventure. And it's our fault? You manage the project better? Right...

The change in account_product_fiscal_classification was a change approved by you even before we approved it, so of course when you approve some change like that upstream you should be ready to deal with following downstream. Yes we need it and we can do that part. And if everybody does the part he needs that would work just fine.

**Comment by rvalyi** [comment]
@mileo @danimar. If you keep contributing constructively in 9.0 we will approve you become the PSC of work you will have done that would be respectively the POS for Brazil and the NFSE. (Also possibly Payroll and Payment gateways if you contribute them in 9.0). That way you will be able to taste how it is to lead a project of your own and standing with the critics and contributors frustrations.

**Comment by danimaribeiro** [comment]
You manage the project better?

Yes I do.

**Comment by rvalyi** [comment]
@danimaribeiro it's easy to claim to do better when you do almost nothing, You are talking about how you help and how we don't. But what is it that you do so much that is helping the project in the OCA branch? You waited until the very last minute to do the branch switch and convinced some people to follow you that way.
This is your right but this isn't helping the project. Now may be you will contribute the NFSE, may be let's judge on actions. In fact to make it clear, I don't put you in the same bag as @mileo who did way more contributions and rant way less.

In general we do not criticize your level of contributions. But sorry pointing the finger at us for your own maintenance trouble when you know exactly why it has been this way and saying we don't help when we do 90% or more of the work was a bit too much to read.

**Comment by danimaribeiro** [comment]
I just see you fixing pep8 issues and writing emails that discourage other people to work on.
I decided not to help because I have to maintain the odoo-brazil, next year I'll start to work again on OCA.

Just another thing, when I sent emails a few months ago asking to flexibilize the process to make the merges faster  I was considered a heretic, and now you guys are doing just that. My point is that you just send emails to prevent other people from contribute, my idea to promote faster development was criticized a lot by you, with a lot of emails, and now you guys just adopted that.

**Comment by renatonlima** [comment]
Hello guys,

@mileo, @danimaribeiro, the main changes of migration that are in the report that you mentioned I had already reported here https://github.com/OCA/l10n-brazil/blob/8.0/l10n_br_account_product/migrations/8.0.2.0.0/pre-migration.py, I can do this script without problems.

But it is important to remember, I as project PSC took much of the work and responsibility in sync branches OCA / 8.0 and contributions made in Odoo-Brazil / 8.0 fork. As most of the work on migrating for the new API, because the fork Odoo-Brazil / 8.0 little was migrated for the new API, most of the major files still uses the old API:

Odoo-Brazil / 8.0 - https://github.com/odoo-brazil/l10n-brazil/blob/8.0/l10n_br_account_product/account_invoice.py
OCA / 8.0 - https://github.com/OCA/l10n-brazil/blob/8.0/l10n_br_account_product/models/account_invoice.py

Odoo-Brazil / 8.0 - https://github.com/odoo-brazil/l10n-brazil/blob/8.0/l10n_br_account_product/l10n_br_account_product.py
OCA / 8.0 - https://github.com/OCA/l10n-brazil/blob/8.0/l10n_br_account_product/models/l10n_br_account_product.py

Odoo-Brazil / 8.0 - https://github.com/odoo-brazil/l10n-brazil/blob/8.0/l10n_br_sale_product/sale.py
OCA / 8.0 - https://github.com/OCA/l10n-brazil/blob/8.0/l10n_br_sale_product/models/sale.py

It's just a comparison so you can see what has been done extensive work to migrate and improve code quality and this will greatly help us in migrating from version 9.0.

To support the migration process and the quality of the project I created tests in the modules of the location and we went from 47% to 64% of coverage.

I did a thorough analysis of your work analyzing 2 months ago analyzing all PRs made in Odoo-Brazil / 8.0 and organized all PRs:

https://github.com/OCA/l10n-brazil/issues?q=is%3Aissue+is%3Aopen+migration+label%3A8.0

Some PRs we integrated the OCA / 8.0 others are open as you can see on the above link.

No one was busier and more than I worked in the migration process and timing of contributions and did it not to impact and work for other contributors.

I do not see the migration scripts as a big problem, I guarantee you that if the teams Aketion, Kmee Trust Code and make a sprint for one day can eliminate this with little effort for everyone.

I do not consider fair after all work that had to appreciate your work and integrate them into the OCA / 8.0 and organize each PR, you do not want to have minimal effort to help synchronize the PRs and generate part of the migration script, as I said I can do this part https://github.com/OCA/l10n-brazil/blob/8.0/l10n_br_account_product/migrations/8.0.2.0.0/pre-migration.py

@mileo you said you are busy lot working in the NT and edoc (The CNAB are new modules that have no impact on l10n_br_core, right?), I'm too busy, I see that this is a problem because we are not working together, I could work on it as well as in other work with greater synergy in the project.

My suggestion is:
1. I do the scripts of the major changes: https://github.com/OCA/l10n-brazil/blob/8.0/l10n_br_account_product/migrations/8.0.2.0.0/pre-migration.py;
2. You and @danimaribeiro make the scripts of the remaining modules;
3. I help and work in the development, testing and approval of NT and refactoring of electronic fiscal documents (edoc);

This is the best way to work together and take advantage of synergies of our works and contributions. This will also help us to maintain the project since it will not have to work on more than one repo.

**Comment by rvalyi** [comment]
@danimaribeiro as with many projects with not enough active experienced contributors we do not always wait 5 days with 2 approvals only. Now, unlike what you suggested, we merge once tests pass. Both on the style and functional parts. And most of the PR's we merged are waiting reviews since June and you kind of waited the last minutes before starting to look at them.
And the fact that we didn't merge before was not forbidding you to put the modules together in some work branch as we do or as Kmee is doing too now and as virtually all OCA project do before everything is merged. Yes I spent time doing pep8. Sorry somebody had to do it and this is sad you considered it a task too dirty for you. But I also always helped Renato design the localization and really pep8 is not all what I do in the localization, please don't try to fool people.

There is a difference between "writing emails to discourage people from contributing" as you say and warning people with no condition to make it to avoid the frustration of failing after many months of efforts. We wouldn't have created the OCA, spent years building the Brazilian Odoo community, made a website that was driving 40% of the traffic to odoo.com if we do not believe in openness as you are accusing us.

The point is Odoo started suboptimal and we know by experience that only exceptionally skilled people would make it, in Brazil and elsewhere, but specially in Brazil where this is harder. The situation being that, Odoo had to evolve and it did at a very fast pace. A pace that is very hard to follow even for us. We are not responsible of this rate of change, but the only option was to follow it until it eventually stabilize. This is not about discouraging people from contributing, this is about avoiding a frustration peak to flame the product before the product is mature. You can see we are one of the OCA localizations the most active with the most contributors so this proves you the grass is not greener on the other side of the fence. Probably the only exception may be Spain where the original localization developers banged the door because of Odoo SA policies and left the community deal with it and yes, also because they have Pedro Baeza which is the OCA alien contributor.

The other day at the Sao Paulo meeting (curious some democratization event we spent time doing, we the little project dictators), there was a guy called Carlos Vercelino (I believe) who testimonied spontaneously and guess what he thanked us so much for having warned him about the pitfalls of using Odoo as an ERP in Brazil. That is he has been using it with success as a RAD framework while he would certainly failed trying to use it as an ERP. So I rather think we save some people from frustrating experience. If you want to let people believe the contrary, you can, but then assume helping their maintenance instead of coming crying at us.

We certainly think Odoo will finally stabilize during versions 9 or 10. And there will be a time you will do apt-get install odoo and have automatic upgrades. We certainly build the OCA for that to happen and unlike what people may think that will not be something bad for us as when this day will happen we will have hundreds more customers or customers hundreds of times bigger. Simply this day has not come yet and certainly we will not accept being pointed as anti-contributions because we don't fall into the easy way of the demagogy.

**Comment by danimaribeiro** [comment]
"I do not see the migration scripts as a big problem, I guarantee you that if the teams Aketion, Kmee Trust Code and make a sprint for one day can eliminate this with little effort for everyone."

That is the kind of reply we would like to see, not this one:

"I think this analysis will be useful for the people using the the temporary branch you mentioned indeed."
"That's not us who promoted this temp branch and I think some people waited a bit too much before making the switch."

@renatonlima We wait saturday to talk about the changes NT, workflow refactory, etc.

**Comment by renatonlima** [comment]
@danimaribeiro

Sorry @danimaribeiro and @mileo, my mother came from Maceio-AL to my house this week and on Saturday I've been a little busy, I did not have time to talk the weekend.

We can talk this week!
